res/controllers/Reloop-Ready-scripts.js

```diff
@@ -0,0 +1,1141 @@
+        shiftOffset: 0x2,
+        input: makeParameterInputHandler(-1),
+    });
+    this.parameterRight = new components.Button({
+        midi: [0x94 + index, 0x29],
+        shiftOffset: 0x2,
+        input: makeParameterInputHandler(1),
+    });
+
+    this.setLoopSizes(this.currentLoopSizeExp);
+};
+
+ReloopReady.LoopRollPadMode.prototype = Object.create(ReloopReady.PadMode.prototype);
+
+
+// Pitch Play mode taken from Roland DJ 505 mapping.
+ReloopReady.Pitch = function(index) {
+    components.ComponentContainer.call(this);
+
+    const PitchPlayRange = {
+        UP: 0,
+        MID: 1,
+        DOWN: 2,
+    };
+
+    //this.ledControl = DJ505.PadMode.SAMPLER;
+    const color = ReloopReady.padColorPalette.Purple;
+    let cuepoint = 1;
+    let range = PitchPlayRange.MID;
+    const theContainer = this;
+
+    this.PerformancePad = function(n) {
+        this.midi = [0x94 + index, 0x14 + n];
+        this.number = n + 1;
+        this.on = ReloopReady.padColorPalette.Purple.dim;
+        this.colorMapper = ReloopReady.padColorMapper;
+        this.colorKey = `hotcue_${this.number}_color`;
+        components.Button.call(this);
+    };
+    this.PerformancePad.prototype = new components.Button({
+        shiftOffset: 8,
+        group: `[Channel${index + 1}]`,
+        outConnect: false,
+        off: ReloopReady.padColorPalette.Off.lit,
+        outputColor: function(colorCode) {
+            // For colored hotcues (shifted only)
+            const midiColor = this.colorMapper.getValueForNearestColor(colorCode);
+            this.send((cuepoint === this.number) ? midiColor : ReloopReady.dimColor(midiColor));
+        },
+        unshift: function() {
+            this.outKey = "pitch_adjust";
+            this.output = function(_value, _group, _control) {
+                let midiColor = color.dim;
+                if ((range === PitchPlayRange.UP && this.number === 5) ||
+                    (range === PitchPlayRange.MID && this.number === 1) ||
+                    (range === PitchPlayRange.DOWN && this.number === 4)) {
+                    midiColor = ReloopReady.padColorPalette.White.lit;
+                }
+                this.send(midiColor);
+            };
+            this.input = function(_channel, _control, value, _status, _group) {
+                const pitchAdjust = (function() {
+                    switch (range) {
+                    case PitchPlayRange.UP:
+                        return this.number + ((this.number <= 4) ? 4 : -5);
+                    case PitchPlayRange.MID:
+                        return this.number - ((this.number <= 4) ? 1 : 9);
+                    case PitchPlayRange.DOWN:
+                        return this.number - ((this.number <= 4) ? 4 : 12);
+                    }
+                }).call(this);
+                engine.setValue(this.group, "pitch_adjust", pitchAdjust);
+                engine.setValue(this.group, `hotcue_${cuepoint}_activate`, value);
```

Comment:
Great that you got the Pitch Play mode incorporated!

After understanding its usage (pitch shift range can be changed by pressing the *PARAMETER 1* buttons; targeted hotcue can be selected while holding *SHIFT*), I noticed that the Pitch Play behavior may be sort of broken when using it with ***loop** hotcues (*Saved Loops*) because of using [`hotcue_${cuepoint}_activate`](https://manual.mixxx.org/2.4/en/chapters/appendix/mixxx_controls#control-[ChannelN]-hotcue_X_activate) here:

**With a *loop hotcue*, selecting a shift will *toggle* the activation state of the saved loop. This means that for an active hotcue loop, the loop will be deactivated, making the track exit the loop. This is probably not the user's expected behavior and makes the usability complicated.**

This can be worked around e.g. by quickly pressing the upcoming next pitch pad twice (double-toggling), but this is quite complicated for well-timed loop pitching.

From the available `hotcue_X_foo` commands on https://manual.mixxx.org/2.4/en/chapters/appendix/mixxx_controls, I didn't find a suitable alternative to `hotcue_${cuepoint}_activate` that would work well with both normal hotcues and loop hotcues.

But if it is possible to first check the kind of the hotcue, then for a hotcue of type *loop hotcue* to use `engine.setValue(this.group, ``hotcue_${cuepoint}_gotoandloop``, value);` would maybe be a viable solution. The loop can still be deactivated on the Reloop Ready quite userfriendly by pressing the conventional hardware *LOOP* rotary encoder.

Reply:
Yikes, that's some unfortunate feature-interaction. wdyt @Holzhaus?

Reply:
A video is worth a thousand words:

[![]](https://github.com/mixxxdj/mixxx/assets/529862/267bfffa-eb75-4a84-bf3f-94a69fa51b49)

Once "Pitch Play mode" is activated, notice how pressing the pad for the selected *Saved Loop hotcue 5" (white color) with the sole intention of changing the pitch also *toggles* the loop activation state (waveform loop background color toggles from whiteish (deactivated) to whiteish-greenish (activated)).
 
`hotcue_${cuepoint}_activate` feels natural per se in *hot cue mode* for the new concept of *Saved Loops* (i.e., `hotcue_${cuepoint}_activate` toggling the loop activation state for a running track).

It's just that `hotcue_${cuepoint}_activate`'s behavior doesn't fit well for *Pitch Play Mode*.

Reply:
prima: with 61ca002628f244983bad4c7b7ef330a1adface9f, I consider the Pitch Play mode issue to be resolved (I think my account doesn't have the authorization to mark *this*/my review as resolved).